File: 1/1000s/14s/1498/NumberOfSubsequencesThatSatisfyTheGivenSumCondition.java
/*
https://leetcode.com/problems/number-of-subsequences-that-satisfy-the-given-sum-condition/description/?envType=daily-question&envId=2025-06-29
*/

class Solution {
    public int numSubseq(int[] nums, int target) {
        Arrays.sort(nums);
        int lengthOfNums=nums.length, modulo=1000000000+7;
        int[] powOfTwo=new int[lengthOfNums];
        powOfTwo[0]=1;
        for(int i=1; i<lengthOfNums; ++i){
            powOfTwo[i]=powOfTwo[i-1]*2;
            powOfTwo[i]%=modulo;
        }
        int left=0, right=lengthOfNums-1, resultantSubsequences=0;
        while(left<=right){
            if(nums[left]+nums[right]<=target){
                resultantSubsequences+=powOfTwo[right-left];
                resultantSubsequences%=modulo;
                left++;
            } else right--;
        }
        return resultantSubsequences;
    }
}

File: 1/1000s/14s/1498/NumberOfSubsequencesThatSatisfyTheGivenSumCondition.py
# https://leetcode.com/problems/number-of-subsequences-that-satisfy-the-given-sum-condition/description/?envType=daily-question&envId=2025-06-29

class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        nums.sort()
        lengthOfNums=len(nums)
        modulo=1000000000+7
        powOfTwo=[]
        for i in range(0, lengthOfNums+1, 1):
            powOfTwo.append(1)
        for i in range(1, lengthOfNums, 1):
            powOfTwo[i]=powOfTwo[i-1]*2
            powOfTwo[i]%=modulo
        left=0
        right=lengthOfNums-1
        resultantSubsequences=0
        while(left<=right):
            if nums[left]+nums[right]<=target:
                resultantSubsequences+=powOfTwo[right-left]
                resultantSubsequences%=modulo
                left+=1
            else:
                right-=1
        return resultantSubsequences